### PR TITLE
refactor(rome_js_formatter): Support arbitrarily deep binary expressions

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

`format_binarish` used to use recursion to format the binary expressions. While this led to somewhat simpler code it had the downside that it stack overflowed for binary expressions with deep left subtrees. 

This PR rewrites the implementation to use an `Iterator` instead that visits the left-hand sides of binary-like-expressions in post order (first visiting the deepest left-hand side, then traversing upwards). 

This PR further fixes a unicode issue in the diff printing logic and renames `operator` to `operator_token` for a consistent naming.

## Test Plan

`cargo test`

Running the formatter on all test repositories (t262, TypeScript...) no longer fails with a stack overflow. There still seems to be some issue that rome enters an infinite loop. I need to look into this separately.

### CODE REVIEW

**Comment by MichaReiser** [comment]
!bench_formatter

File: crates/rome_console/src/diff.rs [hunk 1]
```diff
@@ -154,8 +154,11 @@ impl<'a> Display for Diff<'a> {
                         fmt.write_str(" | ")?;
 
                         let line = change.value().trim_end();
-                        let line_length = line.len().min(MAX_LINE_LENGTH);
-                        let line = &line[..line_length];
+                        let line = line
+                            .char_indices()
+                            .nth(MAX_LINE_LENGTH)
+                            .map(|(byte_index, _)| &line[..byte_index])
+                            .unwrap_or_else(|| line);
```
Comment (MichaReiser) [inline crates/rome_console/src/diff.rs:161]: @leops I believe this still does about the same as you intended, right? It should take the first 250 characters. The only difference now is that it takes the first 250 unicode characters and avoids slicing in the middle of a unicode character.

**Review by leops — APPROVED** [review]
The change in diff printing looks good, I've been meaning to add more tests around the handling of Unicode there anyway. On the `PostorderIterator` I think it looks quite similar to other existing iterators in `rome_rowan`, would it make sense to move this kind of utility there (in a more generic form) ?

**Comment by MichaReiser** [comment]
> The change in diff printing looks good, I've been meaning to add more tests around the handling of Unicode there anyway. On the `PostorderIterator` I think it looks quite similar to other existing iterators in `rome_rowan`, would it make sense to move this kind of utility there (in a more generic form) ?

The `PostorderIterator` is inspired by the `PreorderIterator` and I first planned to write it in a genric way. However, the iterator here is special for two reasons:

* It only traverses into the left-hand sides of binary expressions. It never visits the right-hand side. 
* It only operates on binary expressions. 

Making it more generic would require a lot of additional casting (casting the `SyntaxNode` back to a binary expression) and some way to tell the iterator to only visit Binary like expressions (similar to `skip_subtree`), and skip over the right hand side

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 2]
```diff
@@ -0,0 +1,858 @@
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.next.take()? {
+                VisitEvent::Enter(binary) => {
+                    let left_expression = binary
+                        .left()
+                        .ok()
+                        .and_then(|left| left.as_expression().cloned())
+                        .and_then(|left| JsAnyBinaryLikeExpression::cast(left.syntax().clone()));
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:619]: I haven't thought about implementing `AstNode` to our internal enums! Sometimes I am too narrow minded! 🤯

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 3]
```diff
@@ -0,0 +1,858 @@
...
+
+                    if let Some(expression) = left_expression {
+                        self.next = Some(VisitEvent::Enter(expression));
+                    } else {
+                        // If left is missing or it isn't a binary expression, then format it as part of the parent binary like expression
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:624]:
```suggestion
                        // If left is missing or it isn't a binary like expression, then format it as part of the parent binary like expression
```

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 4]
```diff
@@ -0,0 +1,858 @@
...
+            start: start.syntax().clone(),
+            next: Some(VisitEvent::Enter(start)),
+        }
+    }
+}
+
+impl Iterator for PostorderIterator {
+    type Item = JsAnyBinaryLikeExpression;
+
+    fn next(&mut self) -> Option<Self::Item> {
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:611]: Can explain with a comment what are doing when we call `next`?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 5]
```diff
@@ -0,0 +1,858 @@
...
+#[derive(Debug)]
+enum VisitEvent {
+    Enter(JsAnyBinaryLikeExpression),
+    Exit(JsAnyBinaryLikeExpression),
+}
+
+/// Iterator that visits the left hand sides of all binary like expressions in post-order.
+struct PostorderIterator {
+    next: Option<VisitEvent>,
+    start: JsSyntaxNode,
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:596]: Can we document these two fields?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 6]
```diff
@@ -0,0 +1,858 @@
...
+                VisitEvent::Exit(node) => {
+                    if node.syntax() != &self.start {
+                        self.next = node.syntax().parent().map(|parent| {
+                            let expression = JsAnyBinaryLikeExpression::cast(parent).unwrap();
+                            VisitEvent::Exit(expression)
+                        });
+                    }
+
+                    return Some(node);
+                }
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:637]: Can we document what we are doing here? Also, can you add a `// SAFETY` comment around `unwrap`?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 7]
```diff
@@ -0,0 +1,858 @@
...
+            JsAnyBinaryLikeExpression::JsLogicalExpression(logical) => logical.syntax(),
+            JsAnyBinaryLikeExpression::JsBinaryExpression(binary) => binary.syntax(),
+            JsAnyBinaryLikeExpression::JsInstanceofExpression(instanceof) => instanceof.syntax(),
+            JsAnyBinaryLikeExpression::JsInExpression(in_expression) => in_expression.syntax(),
+        }
+    }
+}
+
+impl JsAnyBinaryLikeExpression {
+    fn can_flatten(&self) -> SyntaxResult<bool> {
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:752]: Can we briefly document when a binary like expression can be flattened?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 8]
```diff
@@ -0,0 +1,858 @@
...
+}
+
+impl ToFormatElement for JsAnyBinaryLikeLeftExpression {
+    fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+        match self {
+            JsAnyBinaryLikeLeftExpression::JsAnyExpression(expression) => {
+                expression.to_format_element(formatter)
+            }
+            JsAnyBinaryLikeLeftExpression::JsPrivateName(private_name) => {
+                private_name.to_format_element(formatter)
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:829]: As both `expression` and `private_name` are typed (not `SyntaxNode`), we have to use `format(formatter)`. `expression` though is a `JsAny` node, so it's fine using `to_format_element`, although `private_name` it is not, so we have to use `format` so we don't bypass the check around comment suppression.

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 9]
```diff
@@ -0,0 +1,858 @@
...
+    Instanceof,
+    In,
+}
+
+/// This function is in charge of formatting a node inside a binaryish expression with parenthesis or not
+///
+/// At the moment this logic is applied only to logical expressions.
+///
+/// A logical expressions should be decorated with parenthesis only if its previous operation has a lower
+/// precedence.
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:162]: I believe this comment is outdated. We now apply parenthesis to all type of binary like expressions. Could you please remove it?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 10]
```diff
@@ -0,0 +1,858 @@
...
+    /// The left hand-side expression and the current operator cannot be flattened.
+    /// Format the left hand side by its own and potentially wrap it in parentheses before formatting
+    /// the right-hand side of the current expression.
+    fn format_new_binary_like_group(
+        &mut self,
+        binary_like_expression: JsAnyBinaryLikeExpression,
+        parent_operator: Option<JsSyntaxToken>,
+        formatter: &Formatter,
+    ) -> FormatResult<()> {
+        if let Some(last) = self.items.last_mut() {
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:350]: As far as I understood `last_mut` is a O(n) operation. Wouldn't it better to keep track of the items pushed and then use this index to surgically retrieve the last item?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 11]
```diff
@@ -0,0 +1,858 @@
...
+        let (trailing_separator, parent_operator_has_comments) = if let Some(parent_operator) =
+            parent_operator.as_ref()
+        {
+            let previous_operator_has_trailing_comments = parent_operator.has_trailing_comments();
+            (
+                if previous_operator_has_trailing_comments {
+                    TrailingTerminator::HardLineBreak
+                } else {
+                    TrailingTerminator::None
+                },
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:390]: nit: a little hard to read, I would move it inside a variable

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 8 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:829]: This is the same as for any other union

https://github.com/rome/tools/blob/437a579e745e4b178a0b3325d412b8e4aa64e3e7/crates/rome_js_formatter/src/js/any/array_assignment_pattern_element.rs#L6-L14

And we call into it using the formatter:

```
let left = parent.left()?;

let formatted = left.format(formatter)?;
```

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:611]: I can extend the documentation on the iterator but next is otherwise documented by the `Iterator` trait.

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 8 above)
Comment (ematipico) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:829]: > This is the same as for any other union

That's what I said. But `JsPrivateName` is not an union, we need to call `format`

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 4 above)
Comment (ematipico) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:611]: Sorry for the misunderstanding, what I meant is **how we are retrieving** `next`, the logic of the algorithm you implemented.

Comment (MichaReiser) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:611]: The struct can describe the overall idea but I rather not document the exact algorithm because it will likely soon be out of sync with the implementation.

Comment (ematipico) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:611]: As this is a complex algorithm, and I don't fully understand what we're doing, I hoped to have some comment as explanation. 

Agree to disagree.

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 8 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:829]: `array_assignment_pattern_element` also calls `to_format_element` for non unions. This is safe because this is only the `to_format_element` dispatch to the right implementation. The `format_binary_like` correctly calls into `.format` for the left hand side above which will handle trivia and such. 

https://github.com/rome/tools/blob/main/crates/rome_js_formatter/src/utils/binary_like_expression.rs#L125-L128

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 10 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:350]: `last_mut` is `O(1)`. It retrieves a reference to the last element. It then mutates the fields in place. An alternative would have been to use `pop` and then push a new element but that wouldn't change anything.
